### PR TITLE
rustc: Move `crate_types` and `stable_crate_id` from `Session` to `GlobalCtxt`

### PR DESCRIPTION
**Pull request by petrochenkov**
Removes two pieces of mutable state.
Follow up to https://github.com/rust-lang/rust/pull/114578.

### CODE REVIEW

**Comment by oli-obk** [comment]
r=me with fallout fixed

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -62,14 +62,15 @@ pub fn ensure_removed(diag_handler: &Handler, path: &Path) {
 /// of the requested outputs for this compilation session.
 pub fn link_binary<'a>(
     sess: &'a Session,
+    crate_types: &[CrateType],
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:65]: Maybe move this into CodegenResults? That will make `-Zlink-only` correctly handle `#![crate_type]` by serializing the crate type in the `.rlink` file.

Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/link.rs]: This looks like the right thing to do, implemented it.
The previous version was pretty mechanical.

**Comment by RalfJung** [comment]
This seems to be part of a general push to remove interior mutability from `Session`? Is the high-level plan/motivation described/tracked anywhere?

**Comment by petrochenkov** [comment]
@RalfJung 
There's not much of a plan, only 3 such fields were discovered in https://github.com/rust-lang/rust/pull/114578, and all of them are already moved.

**Comment by RalfJung** [comment]
Ah I see. :) I was just curious why interior mutability in `Session` is a problem. But anyway, looks like a nice cleanup. :D

**Pull request merged**